File: lw6/Garanin_Timofey/DictionaryComressor/DictionaryComressor/Compressor.cpp
#include "stdafx.h"
#include "Repository.h"
#include "Compressor.h"
#include <cctype>
#include <algorithm>
#include <sstream>
#include <stdlib.h>

CCompressor::CCompressor(size_t order, CMyRepository *repository, std::mutex *mutex)
{
	m_mutex = mutex;
	m_order = order;
	m_repository = repository;
	m_punctuation = { ',', '.', ';', ':', '?', '!', '-', '(', ')', '\'', '\"', ' ' };
	m_specials = { '\n', '\t', '\r', '\f', '\b' };
	m_dictionary = std::make_shared<std::unordered_map<std::string, std::string>>();
	m_textFragmentAfterProcessing = std::make_shared<std::string>();
}

void CCompressor::SetTextFragment(char * text)
{
	m_pTextFragment = text;
}

void CCompressor::SetLengthFragment(size_t length)
{
	m_length = length;
}

void CCompressor::EditFragment()
{
	std::string word;
	if (!m_incompleteWord.empty())
	{
		word = m_incompleteWord;
		m_incompleteWord.clear();
	}
	bool ifPastWasPunctMark = false;
	for (size_t i = 0; i != m_length; ++i)
	{
		if (!IsPunctuaion(m_pTextFragment[i]) && !IsSpecials(m_pTextFragment[i]))
		{
			word += m_pTextFragment[i];
			ifPastWasPunctMark = false;
		}
		else
		{
			if (!ifPastWasPunctMark)
			{
				int number = 0;
				std::transform(word.begin(), word.end(), word.begin(), ::tolower);
				std::stringstream ss;

				m_mutex->lock();
				auto dict = m_repository->GetDictionary();
				if (dict->count(word) > 0)
				{
					
					number = std::stoi(dict->at(word));
					ss << number;
				}
				else
				{
					number = static_cast<int>(dict->size());
					ss << number;
					dict->insert(std::pair<std::string, std::string>(word, std::string(ss.str())));
				}
				m_mutex->unlock();
				
				std::string cv = std::string(ss.str());
				m_textFragmentAfterProcessing->insert(m_textFragmentAfterProcessing->end(), cv.begin(), cv.end());
				if (!IsSpecials(m_pTextFragment[i]))
				{
					m_textFragmentAfterProcessing->insert(m_textFragmentAfterProcessing->end(), m_pTextFragment[i]);
				}
				ifPastWasPunctMark = true;
				word.clear();
			}
			else if (!IsSpecials(m_pTextFragment[i]))
			{
				m_textFragmentAfterProcessing->push_back(m_pTextFragment[i]);
			}
		}
		if (IsSpecials(m_pTextFragment[i]))
		{
			m_textFragmentAfterProcessing->push_back(m_pTextFragment[i]);
			word.clear();
		}
	}
	if (!word.empty())
	{
		m_incompleteWord = std::string(word);
	}
}

size_t CCompressor::GetOrder()
{
	return m_order;
}

std::shared_ptr<std::unordered_map<std::string, std::string>> CCompressor::GetAllDictionary() const
{
	return m_dictionary;
}

std::shared_ptr<std::string> CCompressor::GetAllProcessingText()
{
	return m_textFragmentAfterProcessing;
}

bool CCompressor::IsPunctuaion(char c)
{
	auto index = std::find(m_punctuation.begin(), m_punctuation.end(), c);
	if (index == m_punctuation.end())
	{
		return false;
	}
	return true;
}

bool CCompressor::IsSpecials(char c)
{
	auto index = std::find(m_specials.begin(), m_specials.end(), c);
	if (index == m_specials.end())
	{
		return false;
	}
	return true;
}

File: lw7/Garanin Timofey/DictionaryCompressor/DictionaryCompressor/Compressor.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading;

namespace DictionaryCompressor
{
    
    class Compressor
    {
        private string m_textFragment;
        private string m_incompleteWord;
        private string m_textFragmentAfterProcessing;
        private uint m_length;
        private int m_order;
        private RepositoryProcessedData m_repositoryProcessedData;
        private Dictionary<string, string> m_dictionary;
        private List<char> m_punctuation;
        private List<char> m_specials;
        private Mutex m_mutex;

        public Compressor(int order, RepositoryProcessedData repository, Mutex mutex)
        {
            m_mutex = mutex;
	        m_order = order;
            m_repositoryProcessedData = repository;
            m_punctuation = new List<char>(new char[]{ ',', '.', ';', ':', '?', '!', '-', '(', ')', '\'', '\"', ' ' });
	        m_specials = new List<char>(new char[] { '\n', '\t', '\r', '\f', '\b' });
	        m_dictionary = new Dictionary<string, string>();
            m_textFragmentAfterProcessing = "";

        }

        public void EditFragment()
        {
            string word = "";
            if (!String.IsNullOrEmpty(m_incompleteWord))
            {
                word = m_incompleteWord;
                m_incompleteWord.Remove(0);
            }
            bool ifPastWasPunctMark = false;
            for (int i = 0; i != m_textFragment.Length; ++i)
            {
                
                if (!IsPunctuaion(m_textFragment.ElementAt(i)) && !IsSpecials(m_textFragment.ElementAt(i)))
                {
                    var tt = m_textFragment.ElementAt(i).ToString();
                    word = word.Insert(word.Length, tt);
                    ifPastWasPunctMark = false;
                }
                else
                {
                    if (!ifPastWasPunctMark)
                    {
                        m_mutex.WaitOne() ;
                        string value = "";
                        var dict = m_repositoryProcessedData.GetDictionary();
                        if (!dict.TryGetValue(word, out value))
                        {
                            value = Convert.ToString(dict.Count);
                            dict.Add(word, value);
                        }
                        m_mutex.ReleaseMutex();
                        m_textFragmentAfterProcessing = m_textFragmentAfterProcessing.Insert(m_textFragmentAfterProcessing.Length, value);
				        if (!IsSpecials(m_textFragment.ElementAt(i)))
				        {
                            m_textFragmentAfterProcessing = m_textFragmentAfterProcessing.Insert(m_textFragmentAfterProcessing.Length, m_textFragment.ElementAt(i).ToString());
				        }
                        ifPastWasPunctMark = true;
                        word = word.Remove(0, word.Length);
                    }
			        else if (!IsSpecials(m_textFragment.ElementAt(i)))
			        {
                        m_textFragmentAfterProcessing = m_textFragmentAfterProcessing.Insert(m_textFragmentAfterProcessing.Length, m_textFragment.ElementAt(i).ToString());
			        }
		        }
		        if (IsSpecials(m_textFragment.ElementAt(i)))
		        {
                    m_textFragmentAfterProcessing = m_textFragmentAfterProcessing.Insert(m_textFragmentAfterProcessing.Length, m_textFragment.ElementAt(i).ToString());
                    word = word.Remove(0, word.Length);
                }
	        }
	        if (!String.IsNullOrEmpty(word))
	        {
		        m_incompleteWord = word;
	        }
        }

        public void SetTextFragment(string text)
        {
            m_textFragment = text;
        }

        public void SetLengthFragment(uint length)
        {
            m_length = length;
        }

        public int GetOrder()
        {
            return m_order;
        }

        public Dictionary<string, string> GetAllDictionary()
        {
	        return m_dictionary;
        }

        public string GetAllProcessingText()
        {
            return m_textFragmentAfterProcessing;
        }

        private bool IsPunctuaion(char c)
        {
            var index = m_punctuation.IndexOf(c);
            if (index == -1)
            {
                return false;
            }
            return true;
        }

        private bool IsSpecials(char c)
        {
            var index = m_specials.IndexOf(c);
            if (index == -1)
            {
                return false;
            }
            return true;
        }

    }
}
